### PR TITLE
Version 3.0.0

### LINKED ISSUE #17: Concept for fully randomized overworld items
I mentioned this in the Discord, but I want to jot it down here as well so it doesn't get lost to the sands of time.

As I understand it, the reason progressive items have their own special "pick from a list"-style randomization is that the mod doesn't see _item locations_, it only sees _item IDs_. Since most overworld item locations contain things like Titanite Shards that aren't at all unique, the mod isn't able to distinguish whether the player picked up a Titanite Shard in High Wall of Lothric or bought one in Firelink Shrine.

But I think this issue could be solved without needing to access additional data in the mod itself. The parameter file and map formats are well-understood, so—before booting the game—you could edit the params to add a few hundred additional fake items that exist only to have unique IDs, and then edit the maps to point all the overworld progressive pickups at those fake items. Then you could have an option on the randomizer server to track whether progressive items have been faked out and add them to the item pool if so.

### PR DESCRIPTION
**Pull request by nex3**
This PR fully implements the changes discussed in #17, integrating this client with [a branch] of @thefifthmatt's static randomizer so that _all_ item locations can be randomized rather than just unique items.

[a branch]: https://github.com/nex3/SoulsRandomizers/tree/archipelago-server

Major functionality changes:

* The static randomizer now replaces all in-world item pickups with synthetic items whose paramdata encodes the Archipelago location ID and (for local items) the DS3 item ID. Specifically:

  * When a local item is picked up, this client decodes the paramdata, tells Archipelago that the corresponding location was accessed, and replaces the synthetic item with the original DS3 item before it enters the player's inventory.
  * When a foreign item is picked up, the client decodes the paramdata, tells Archipelago that the corresponding location was accessed, and deletes the item from the player's inventory.

* Uses ModEngine2 rather than manual dinput8.dll injection to be compatible with other mods (particularly the static randomizer). This also makes it compatible with DS3 1.15.2.

* More gracefully handles connection and disconnection:

   * Shows a banner in-game when a connection is established.
   * If foreign items are picked up while disconnected, they will be sent once reconnected.
   * More robust checks for when the player is fully loaded into the game.

* Drops progressive item logic since that's no longer necessary to handle non-unique items.

* Supports granting Path of the Dragon through a custom event created by the static randomizer.

Major infrastructure changes:

* Uses vcpkg for dependencies where possible (similar to #27).
* Uses MinHook to hook into internal DS3 functionality with AOBs, adding more version flexibility.
* Avoids assembly in favor of explicit C code, manipulating defined structs where possible.
* Adds a GitHub Actions workflow to automatically build and publish tagged releases.
* Uses `spdlog` for compatibility with ModEngine2.

Instructions for building this client are in the README. You can test it by taking [the latest beta release](https://github.com/nex3/Dark-Souls-III-Archipelago-client/releases) and overwriting its `archipelago.dll` with the one you built locally.

Closes #17

### CODE REVIEW

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+    - name: Download ModEngine2
+      run: |
+        Invoke-WebRequest -OutFile ME2.zip -Uri 'https://github.com/soulsmods/ModEngine2/releases/download/release-2.1.0/ModEngine-2.1.0.0-win64.zip'
+        Expand-Archive ME2.zip -DestinationPath .
+
+    - name: Download Archipelago repo
+      run: |
+        Invoke-WebRequest -OutFile Archipelago.zip -Uri "https://github.com/nex3/Archipelago/archive/refs/tags/ds3-${{ needs.version.outputs.version }}.zip"
+        Expand-Archive Archipelago.zip .
```
Comment (Marechal-L) [inline .github/workflows/release.yaml:171]: Can we remove this from the final version once the server and the client has been merged ?

File: archipelago-client/ArchipelagoInterface.cpp [hunk 2]
```diff
@@ -67,14 +73,21 @@ BOOL CArchipelago::Initialise(std::string URI) {
...
 
+		Core->connected = true;
+		ItemRandomiser->sendMissedItems();
+
 		});
 	ap->set_slot_disconnected_handler([]() {
-		Core->Logger("Slot disconnected");
+		spdlog::info("Slot disconnected");
+		Core->connected = false;
+		GameHook->showBanner(L"Archipelago disconnected! Don't pick up any items until it reconnects");
```
Comment (Marechal-L) [inline archipelago-client/ArchipelagoInterface.cpp:83]: Maybe this behavior belongs also to set_socket_disconnected_handler, I need to give it a try

File: release/README.txt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+* `enable_boss_locations` is now controlled by the `soul_locations` option.
+
+* `enable_progressive_locations` was removed because all locations are now individually randomized rather than replaced with a progressive list.
+
+* `pool_type` has been removed. Since there are no longer any non-randomized items in randomized categories, there's not a meaningful distinction between "shuffle" and "various" mode.
+
+In addition, the following options have changed:
+
+* `enable_*_locations` options have all been deprecated. Instead, there are a number of named location groups that you can add to the `exclude_locations` option to prevent them from containing important items. These location groups are:
```
Comment (Marechal-L) [inline release/README.txt]: Just like in the AP server, I suggest to mark those locations as removed while keeping the explainations on how to setup "exclude_locations" properly

**Comment by Marechal-L** [comment]
I don't encountered any major issues during my testings and the result looks amazing! 

I still have two minor issues and I would be glad to investigate on my own to fix this but I would like your feedback first : 

1. Items injected by the SoulsRandomizers are displayed with the default label "?good's name?" when the game is played in French ( and probably any language but English ). I didn't dig into the SoulsRandomizers yet but if you have any idea to fix this it would be great.

2. Local items purchased from the shop do not trigger a check to the server but world items do ? I'll probably need to do more testing to fully understand your logic behind it but if you already have an explanation for me I would be greateful

File: .github/workflows/release.yaml (hunk 1 above)
Comment (nex3) [reply .github/workflows/release.yaml:171]: Yes. Even before then this should _really_ only include the apworld, but people have to use my fork of the entire repo until https://github.com/ArchipelagoMW/Archipelago/pull/2403 is merged.

**Comment by nex3** [comment]
> Items injected by the SoulsRandomizers are displayed with the default label "?good's name?" when the game is played in French ( and probably any language but English ). I didn't dig into the SoulsRandomizers yet but if you have any idea to fix this it would be great.

This is currently a restriction intrinsic to the static randomizer—it only overwrites the messages for the `engus` localization. Hopefully we can talk to Matt about making that more flexible, but I think it may have to be a post-3.0.0 thing.

> Local items purchased from the shop do not trigger a check to the server but world items do ? I'll probably need to do more testing to fully understand your logic behind it but if you already have an explanation for me I would be greateful

Yeah, this is definitely not ideal but it's _extremely_ hard to fix. The core of the issue is, in order to make DS3-native shop items look right (show up in the right tabs, in the right order, with the right stats) we need them to be _real_ items rather than the synthetic fakes we use for other drops. But since they're real items, we can't embed extra metadata to figure out which location they represent once you purchase them.

It's possible we could work around this at least to some degree if we could get a listing of all the items that exist in the shop and use that to determine which location was bought... although even then if you have, for example, six different Embers available in the Firelink shop from different locations it's not clear how you'd figure out which Ember came from which source.

**Pull request closed without merging**